File: app/src/main/java/lt/bulevicius/tessonetapp/app/NetworkModule.java
package lt.bulevicius.tessonetapp.app;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import lt.bulevicius.tessonetapp.BuildConfig;
import lt.bulevicius.tessonetapp.app.utils.NetworkSchedulerProvider;
import lt.bulevicius.tessonetapp.app.utils.SchedulerProvider;
import lt.bulevicius.tessonetapp.network.TessonetApi;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import timber.log.Timber;

import static okhttp3.logging.HttpLoggingInterceptor.Level.BODY;

@Module
final class NetworkModule {

    @Provides
    @Singleton
    final Retrofit retrofit(OkHttpClient client) {
        return new Retrofit.Builder().addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                                     .addConverterFactory(GsonConverterFactory.create())
                                     .client(client)
                                     .baseUrl(BuildConfig.API)
                                     .build();
    }

    @Provides
    @Singleton
    final TessonetApi api(Retrofit retrofit) {
        return retrofit.create(TessonetApi.class);
    }

    @Provides
    @Singleton
    final OkHttpClient okHttpClient(HttpLoggingInterceptor loggingInterceptor) {
        return new OkHttpClient.Builder().addInterceptor(loggingInterceptor).build();
    }

    @Provides
    @Singleton
    final HttpLoggingInterceptor loggingInterceptor() {
        return new HttpLoggingInterceptor(message -> Timber.d(message)).setLevel(BODY);
    }

    @Provides
    @Singleton
    final SchedulerProvider networkSchedulerProvider() {
        return new NetworkSchedulerProvider(Schedulers.io(), AndroidSchedulers.mainThread());
    }
}

File: app/src/main/java/com/edvinas/balkaitis/party/utils/network/NetworkModule.kt
package com.edvinas.balkaitis.party.utils.network

import com.edvinas.balkaitis.party.BuildConfig
import com.edvinas.balkaitis.party.login.network.LoginService
import com.edvinas.balkaitis.party.servers.network.ServersService
import com.edvinas.balkaitis.party.utils.schedulers.Io
import dagger.Module
import dagger.Provides
import io.reactivex.Scheduler
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import javax.inject.Singleton

@Module
abstract class NetworkModule {

    @Module
    companion object {
        private const val AUTHORIZATION_HEADER = "orization"
        private const val BEARER = "Bearer"


        @JvmStatic
        @Singleton
        @Provides
        fun provideRetrofit(client: OkHttpClient, @Io scheduler: Scheduler): Retrofit {
            return Retrofit.Builder()
                .baseUrl(BuildConfig.BASE_URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.createWithScheduler(scheduler))
                .build()
        }

        @JvmStatic
        @Provides
        fun provideOkHttpClient(): OkHttpClient {
            return OkHttpClient.Builder()
                .addInterceptor(HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY))
                .build()
        }

        @JvmStatic
        @Provides
        fun provideLoginService(retrofit: Retrofit): LoginService {
            return retrofit.create(LoginService::class.java)
        }

        @JvmStatic
        @Provides
        fun provideServersService(retrofit: Retrofit): ServersService {
            return retrofit.create(ServersService::class.java)
        }
    }
}
